File: src/main/scala/io/github/jlprat/akka/http/workshop/bookstore/model/Review.scala
package io.github.jlprat.akka.http.workshop.bookstore.model

import java.util.Date

import io.github.jlprat.akka.http.workshop.bookstore.model.Review.Stars

object Review {

  /**
    * Models the star rating system
    */
  case class Stars protected[bookstore](stars: String)
  val `*` = Stars("1")
  val `**` = Stars("2")
  val `***` = Stars("3")
  val `****` = Stars("4")
  val `*****` = Stars("5")

  object Stars {
    def fromString(stars: String): Stars = stars match {
      case "1" => `*`
      case "2" => `**`
      case "3" => `***`
      case "4" => `****`
      case "5" => `*****`
    }
  }
}

case class Review(author: Author, date: Long = new Date().getTime, comment: String, stars: Stars)

File: src/main/java/io/github/jlprat/akka/http/workshop/java/bookstore/model/Review.java
package io.github.jlprat.akka.http.workshop.java.bookstore.model;

/**
 * Created by jlprat on 23.04.17.
 */
public class Review {

    private Author author;
    private String comment;
    private long timestamp;
    private int stars;

    public Review(Author author, String comment, int stars) {
        this.author = author;
        this.comment = comment;
        this.stars = stars;
        this.timestamp = System.currentTimeMillis();
    }

    public Author getAuthor() {
        return author;
    }

    public void setAuthor(Author author) {
        this.author = author;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    public int getStars() {
        return stars;
    }

    public void setStars(int stars) {
        this.stars = stars;
    }

    @Override
    public String toString() {
        return "Review{" +
                "author=" + author +
                ", comment='" + comment + '\'' +
                ", timestamp=" + timestamp +
                ", stars=" + stars +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Review review = (Review) o;

        if (timestamp != review.timestamp) return false;
        if (stars != review.stars) return false;
        if (author != null ? !author.equals(review.author) : review.author != null) return false;
        return comment != null ? comment.equals(review.comment) : review.comment == null;
    }

    @Override
    public int hashCode() {
        int result = author != null ? author.hashCode() : 0;
        result = 31 * result + (comment != null ? comment.hashCode() : 0);
        result = 31 * result + (int) (timestamp ^ (timestamp >>> 32));
        result = 31 * result + stars;
        return result;
    }
}
